streams/client_test.go

```diff
@@ -210,143 +210,133 @@ func TestPublishEvents(t *testing.T) {
-		putRecordsErr := []error{nil}
-		client.streams = &StubClient{out: putRecordsOut, err: putRecordsErr}
-
-		rest, err := client.publishEvents(events)
-		if err != nil {
-			t.Errorf("unexpected error: %v", err)
-		}
-		if len(rest) != 0 {
-			t.Errorf("unexpected number of remaining events: %d", len(rest))
-		}
-	}
-
-	{
-		// Kinesis received the event but it was not persisted, probably due to underlying infrastructure failure
-		codecData := [][]byte{[]byte("boom")}
-		codecErr := []error{nil}
-		client.encoder = &StubCodec{dat: codecData, err: codecErr}
-
-		putRecordsOut := []*kinesis.PutRecordsOutput{
-			&kinesis.PutRecordsOutput{
-				Records: []*kinesis.PutRecordsResultEntry{
-					&kinesis.PutRecordsResultEntry{
-						ErrorCode: aws.String("simulated_error"),
+			putRecordsErr:   []error{nil},
+			codecData:       [][]byte{[]byte("boom")},
+			codecErr:        []error{nil},
+			eventsToPublish: []publisher.Event{event},
+			remainingEvents: 0,
+		},
+		{ // Kinesis received the event but it was not persisted, probably due to underlying infrastructure failure
+			testname: "event not persisted",
+			putRecordsOut: []*kinesis.PutRecordsOutput{
+				{
+					Records: []*kinesis.PutRecordsResultEntry{
+						{
+							ErrorCode: aws.String("simulated_error"),
+						},
 					},
+					FailedRecordCount: aws.Int64(1),
 				},
-				FailedRecordCount: aws.Int64(1),
 			},
-		}
-		putRecordsErr := []error{nil}
-		client.streams = &StubClient{out: putRecordsOut, err: putRecordsErr}
-
-		rest, err := client.publishEvents(events)
-		if err != nil {
-			t.Errorf("unexpected error: %v", err)
-		}
-		if len(rest) != 1 {
-			t.Errorf("unexpected number of remaining events: %d", len(rest))
-		}
+			putRecordsErr:   []error{nil},
+			codecData:       [][]byte{[]byte("boom")},
+			codecErr:        []error{nil},
+			eventsToPublish: []publisher.Event{event},
+			remainingEvents: 1,
+		},
+		{ // Kinesis received the event but it was not persisted, due to throughput exceeded error
+			testname: "throughput exceeded",
+			putRecordsOut: []*kinesis.PutRecordsOutput{
+				{
+					Records: []*kinesis.PutRecordsResultEntry{
+						{
+							ErrorCode: aws.String("ProvisionedThroughputExceededException"),
+						},
+					},
+					FailedRecordCount: aws.Int64(1),
+				},
+			},
+			putRecordsErr:   []error{nil},
+			codecData:       [][]byte{[]byte("boom")},
+			codecErr:        []error{nil},
+			eventsToPublish: []publisher.Event{event},
+			remainingEvents: 1,
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.testname, func(t *testing.T) {
+			client.encoder = &StubCodec{dat: tc.codecData, err: tc.codecErr}
+			client.streams = &StubClient{out: tc.putRecordsOut, err: tc.putRecordsErr}
+			rest, err := client.publishEvents(tc.eventsToPublish)
+			if err != nil {
+				t.Errorf("unexpected error: %+v", err)
+			}
+			if len(rest) != tc.remainingEvents {
+				t.Errorf("unexpected number of remaining events got:%d want:%d", len(rest), tc.remainingEvents)
+			}
```

Comment:
Where do we test that there is a retry for ThroughputExceeded?

Reply:
testname: "throughput exceeded",
expectedRemainingEvents: 1,
this means that when there is an error the event is send back to retry